File: ado/PgWireCoreClient/PgWireAdo/utils/DataMessage.cs
using System;
using System.Buffers.Binary;
using System.Text;

namespace PgWireAdo.utils;

public class DataMessage
{
    protected bool Equals(DataMessage other)
    {
        return _id.Equals(other._id) ;
    }

    public override bool Equals(object? obj)
    {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        if (obj.GetType() != this.GetType()) return false;
        return Equals((DataMessage)obj);
    }

    public override int GetHashCode()
    {
        return HashCode.Combine(_id);
    }

    private Guid _id = Guid.NewGuid();

    public Guid Id => _id;

    public long Timestamp { get; }
    private int _cursor = 0;
    public char Type { get; }
    public int Length { get; }
    public byte[] Data { get; }


    public DataMessage(char type, int length, byte[] data,long timestap)
    {
        Timestamp = timestap;
        Type = type;
        Length = length;
        Data = data;
    }
    

    public int ReadInt()
    {
        var result = BinaryPrimitives.ReverseEndianness(BitConverter.ToInt32(Data, _cursor));
        _cursor += 4;
        return result;
    }

    public int ReadInt32()
    {
        return ReadInt();
    }

    public short GetShort()
    {
        var intArray = new byte[2];
        var result = BinaryPrimitives.ReverseEndianness(BitConverter.ToInt16(Data, _cursor));
        _cursor += 2;
        return result;
    }

    public String ReadAsciiString()
    {
        var ms = new MemoryStream();

        var count = 0;
        var start = _cursor;
        for (; _cursor < Data.Length; _cursor++)
        {

            if (Data[_cursor] == 0x00)
            {
                _cursor++;
                break;
            }
            ms.WriteByte(Data[_cursor]);
            count++;
        }

        return ASCIIEncoding.Default.GetString(ms.ToArray());
    }

    public String ReadUTF8String()
    {
        var ms = new MemoryStream();

        var count = 0;
        var start = _cursor;
        for (; _cursor < Data.Length; _cursor++)
        {

            if (Data[_cursor] == 0x00)
            {
                _cursor++;
                break;
            }
            ms.WriteByte(Data[_cursor]);
            count++;
        }

        return UTF8Encoding.Default.GetString(ms.ToArray());
    }

    public byte[] ReadBytes(int parameterLength)
    {
        var result = new byte[parameterLength];
        for (var i = 0; i < parameterLength; i++, _cursor++)
        {
            result[i] = Data[_cursor];
        }
        return result;
    }

    public byte ReadByte()
    {
        var result = Data[_cursor];
        _cursor++;
        return result;
    }

    public short ReadInt16()
    {
        return GetShort();
    }
}

File: pgwire-jdbc-server/pgwire-jdbc-bridge/src/main/java/org/kendar/pgwire/commons/DataMessage.java
package org.kendar.pgwire.commons;

import java.nio.ByteBuffer;

public class DataMessage {
    private int cursor = 0;
    private char type;
    private int length;
    private byte[] data;

    public char getType() {
        return type;
    }

    public void setType(char type) {
        this.type = type;
    }

    public int getLength() {
        return length;
    }

    public void setLength(int length) {
        this.length = length;
    }

    public byte[] getData() {
        return data;
    }

    public void setData(byte[] data) {
        this.data = data;
    }

    public DataMessage(char type, int length, byte[] data) {
        this.type = type;
        this.length = length;
        this.data = data;
    }

    public int readInt(){
        var result = ByteBuffer.wrap(data,cursor,4).getInt();
        cursor+=4;
        return result;
    }

    public short getShort() {
        var result = ByteBuffer.wrap(data,cursor,2).getShort();
        cursor+=2;
        return result;
    }

    public String readString(){
        var base = cursor;
        var count =0;
        for(;cursor<data.length;cursor++){

            if(data[cursor]==0x00){
                cursor++;
                break;
            }
            count++;
        }
        return new String(data,base,count);
    }

    public byte[] readBytes(int parameterLength) {
        var result = new byte[parameterLength];
        for(var i=0;i<parameterLength && cursor<data.length;i++,cursor++){
            result[i]=data[cursor];
        }
        return result;
    }

    public byte readByte() {
        cursor++;
        return data[cursor-1];
    }
}
